# Profitable Apps Dataset
Data analyst project for a company which creates Android and iOS mobile apps, available on the Google Play and App Store. The company only builds apps that are free to download and install, and their main source of revenue consists of in-app ads. 

This means that the number of users of apps determines the revenue for any given app — the more users who see and engage with the ads, the better. The goal for this project is to analyze data to help the developers understand what type of apps are likely to attract more users.

In [1]:
import numpy as np
import pandas as pd

## Opening & Exploring the Dataset

Our dataset contains a sample of app data. **apple_data** contains data on 7200 apps, **google_data** contains data on 9660 apps.

Apple Store data available from: https://www.kaggle.com/datasets/ramamet4/app-store-apple-data-set-10k-apps

Google Play Store data available from: https://www.kaggle.com/datasets/lava18/google-play-store-apps

In [2]:
apple_data = pd.read_csv('AppleStore.csv')
google_data = pd.read_csv('googleplaystore.csv')

In [3]:
apple_data

id                                    track_name  size_bytes  \
0      284882215                                      Facebook   389879808   
1      389801252                                     Instagram   113954816   
2      529479190                                Clash of Clans   116476928   
3      420009108                                    Temple Run    65921024   
4      284035177                       Pandora - Music & Radio   130242560   
...          ...                                           ...         ...   
7192  1170406182  Shark Boom - Challenge Friends with your Pet   245415936   
7193  1069830936                           【謎解き】ヤミすぎ彼女からのメッセージ    16808960   
7194  1070052833                                    Go!Go!Cat!    91468800   
7195  1081295232   Suppin Detective: Expose their true visage!    83026944   
7196   977965019           みんなのお弁当 by クックパッド ~お弁当をレシピ付きで記録・共有~    51174400   

     currency  price  rating_count_tot  rating_count_ver  user_rating  \
0         USD    0.0           2974676               212          3.5   
1         USD    0.0           2161558              1289          4.5   
2         USD    0.0           2130805               579          4.5   
3         USD    0.0           1724546              3842          4.5   
4         USD    0.0           1126879              3594          4.0   
...       ...    ...               ...               ...          ...   
7192      USD    0.0                 0                 0          0.0   
7193      USD    0.0                 0                 0          0.0   
7194      USD    0.0                 0                 0          0.0   
7195      USD    0.0                 0                 0          0.0   
7196      USD    0.0                 0                 0          0.0   

      user_rating_ver      ver cont_rating        prime_genre  \
0                 3.5     95.0          4+  Social Networking   
1                 4.0    10.23         12+      Photo & Video   
2                 4.5  9.24.12          9+              Games   
3                 4.0    1.6.2          9+              Games   
4                 4.5    8.4.1         12+              Music   
...               ...      ...         ...                ...   
7192              0.0    1.0.9          4+              Games   
7193              0.0      1.2          9+               Book   
7194              0.0    1.1.2         12+              Games   
7195              0.0    1.0.3         12+      Entertainment   
7196              0.0    1.4.0          4+       Food & Drink   

      sup_devices.num  ipadSc_urls.num  lang.num  vpp_lic  
0                  37                1        29        1  
1                  37                0        29        1  
2                  38                5        18        1  
3                  40                5         1        1  
4                  37                4         1        1  
...               ...              ...       ...      ...  
7192               38                5         1        1  
7193               38                0         1        1  
7194               37                2         2        1  
7195               40                0         1        1  
7196               37                0         1        1  

[7197 rows x 16 columns]

In [4]:
google_data

App             Category  \
0         Photo Editor & Candy Camera & Grid & ScrapBook       ART_AND_DESIGN   
1                                    Coloring book moana       ART_AND_DESIGN   
2      U Launcher Lite – FREE Live Cool Themes, Hide ...       ART_AND_DESIGN   
3                                  Sketch - Draw & Paint       ART_AND_DESIGN   
4                  Pixel Draw - Number Art Coloring Book       ART_AND_DESIGN   
...                                                  ...                  ...   
10836                                   Sya9a Maroc - FR               FAMILY   
10837                   Fr. Mike Schmitz Audio Teachings               FAMILY   
10838                             Parkinson Exercices FR              MEDICAL   
10839                      The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
10840      iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

       Rating Reviews                Size     Installs  Type Price  \
0         4.1     159                 19M      10,000+  Free     0   
1         3.9     967                 14M     500,000+  Free     0   
2         4.7   87510                8.7M   5,000,000+  Free     0   
3         4.5  215644                 25M  50,000,000+  Free     0   
4         4.3     967                2.8M     100,000+  Free     0   
...       ...     ...                 ...          ...   ...   ...   
10836     4.5      38                 53M       5,000+  Free     0   
10837     5.0       4                3.6M         100+  Free     0   
10838     NaN       3                9.5M       1,000+  Free     0   
10839     4.5     114  Varies with device       1,000+  Free     0   
10840     4.5  398307                 19M  10,000,000+  Free     0   

      Content Rating                     Genres      Last Updated  \
0           Everyone               Art & Design   January 7, 2018   
1           Everyone  Art & Design;Pretend Play  January 15, 2018   
2           Everyone               Art & Design    August 1, 2018   
3               Teen               Art & Design      June 8, 2018   
4           Everyone    Art & Design;Creativity     June 20, 2018   
...              ...                        ...               ...   
10836       Everyone                  Education     July 25, 2017   
10837       Everyone                  Education      July 6, 2018   
10838       Everyone                    Medical  January 20, 2017   
10839     Mature 17+          Books & Reference  January 19, 2015   
10840       Everyone                  Lifestyle     July 25, 2018   

              Current Ver         Android Ver  
0                   1.0.0        4.0.3 and up  
1                   2.0.0        4.0.3 and up  
2                   1.2.4        4.0.3 and up  
3      Varies with device          4.2 and up  
4                     1.1          4.4 and up  
...                   ...                 ...  
10836                1.48          4.1 and up  
10837                 1.0          4.1 and up  
10838                 1.0          2.2 and up  
10839  Varies with device  Varies with device  
10840  Varies with device  Varies with device  

[10841 rows x 13 columns]

Note that the Apple dataset has **16** columns and the Google dataset has **13** columns.

In [5]:
# Apple column values:
for col in apple_data.columns:
    print(col)

id
track_name
size_bytes
currency
price
rating_count_tot
rating_count_ver
user_rating
user_rating_ver
ver
cont_rating
prime_genre
sup_devices.num
ipadSc_urls.num
lang.num
vpp_lic


In [6]:
# Google column values:
for col in google_data.columns:
    print(col)

App
Category
Rating
Reviews
Size
Installs
Type
Price
Content Rating
Genres
Last Updated
Current Ver
Android Ver


## Cleaning the Dataset
Before we begin analzying the data, we will want to ensure that the values are correct, and that there are no missing or null values

### Deleting Incorrect Data
The Google Play Store dataset has a discussion section, which outlines an error in the dataset in row 10472. 

In [7]:
# Correct row:
print(google_data.loc[1])
print()
# Row with error:
print(google_data.loc[10472])

App                     Coloring book moana
Category                     ART_AND_DESIGN
Rating                                  3.9
Reviews                                 967
Size                                    14M
Installs                           500,000+
Type                                   Free
Price                                     0
Content Rating                     Everyone
Genres            Art & Design;Pretend Play
Last Updated               January 15, 2018
Current Ver                           2.0.0
Android Ver                    4.0.3 and up
Name: 1, dtype: object

App               Life Made WI-Fi Touchscreen Photo Frame
Category                                              1.9
Rating                                                 19
Reviews                                              3.0M
Size                                               1,000+
Installs                                             Free
Type                                                    

The rating value of 19 is off (ratings are between 1-5), the Content Rating is missing, and the Price value is incorrect.

We can delete this row and check for other null/missing values in the dataset.

In [8]:
print(len(google_data))
google_data.drop(10472, inplace=True)
print(len(google_data))

10841
10840


## Checking for Null Values
Let's check the total number of null (empty) values for each dataset:

In [9]:
apple_data.isnull().sum()

id                  0
track_name          0
size_bytes          0
currency            0
price               0
rating_count_tot    0
rating_count_ver    0
user_rating         0
user_rating_ver     0
ver                 0
cont_rating         0
prime_genre         0
sup_devices.num     0
ipadSc_urls.num     0
lang.num            0
vpp_lic             0
dtype: int64

No null values in app_data, checking google_data:

In [10]:
google_data.isnull().sum()

App                  0
Category             0
Rating            1474
Reviews              0
Size                 0
Installs             0
Type                 1
Price                0
Content Rating       0
Genres               0
Last Updated         0
Current Ver          8
Android Ver          2
dtype: int64

1,474 apps have no value for the Rating Category, 1 has no value in Type, 8 have no value for Current Ver and 2 have no value for Android Ver.

Which rows should be removed? We will leave the apps with null values in columns we aren't concerned with (Current ver & Android ver).

##### Null Values in Type

In [11]:
google_data['Type'].value_counts(dropna=False)

Free    10039
Paid      800
NaN         1
Name: Type, dtype: int64

We are only interested in Free Apps, so an app with no value in the Type column is bad for our dataset. Let's look into the app with the null value:

In [12]:
google_data[google_data['Type'].isnull()]

App Category  Rating Reviews                Size  \
9148  Command & Conquer: Rivals   FAMILY     NaN       0  Varies with device   

     Installs Type Price Content Rating    Genres   Last Updated  \
9148        0  NaN     0   Everyone 10+  Strategy  June 28, 2018   

             Current Ver         Android Ver  
9148  Varies with device  Varies with device

The app has a NaN value under Type but a price of 0 -- instead of deleting it we can update it to 'Free':

In [13]:
google_data.loc[9148, 'Type'] = 'Free'

In [14]:
google_data['Type'].value_counts(dropna=False)

Free    10040
Paid      800
Name: Type, dtype: int64

##### Null values in Rating

In [15]:
google_data.Rating.unique()

array([4.1, 3.9, 4.7, 4.5, 4.3, 4.4, 3.8, 4.2, 4.6, 3.2, 4. , nan, 4.8,
       4.9, 3.6, 3.7, 3.3, 3.4, 3.5, 3.1, 5. , 2.6, 3. , 1.9, 2.5, 2.8,
       2.7, 1. , 2.9, 2.3, 2.2, 1.7, 2. , 1.8, 2.4, 1.6, 2.1, 1.4, 1.5,
       1.2])

We might not need rating data for our purposes, but we don't want to leave it empty. Instead, we can change the null values to 'Unrated':

In [16]:
google_data['Rating'].fillna("Unrated", inplace=True)

In [17]:
google_data.Rating.unique()

array([4.1, 3.9, 4.7, 4.5, 4.3, 4.4, 3.8, 4.2, 4.6, 3.2, 4.0, 'Unrated',
       4.8, 4.9, 3.6, 3.7, 3.3, 3.4, 3.5, 3.1, 5.0, 2.6, 3.0, 1.9, 2.5,
       2.8, 2.7, 1.0, 2.9, 2.3, 2.2, 1.7, 2.0, 1.8, 2.4, 1.6, 2.1, 1.4,
       1.5, 1.2], dtype=object)

## Duplicate Values
We want to ensure that the data in our dataset is not only correct and non-empty, but also that there are no duplicate entries.

Let's look at the title of apps to see if an app is listed multiple times:

In [18]:
unique_apple_apps = apple_data['track_name'].nunique()
unique_google_apps = google_data['App'].nunique()

print("Unique Apple apps:", unique_apple_apps)
print("Total Apple apps:", len(apple_data))
print("Duplicate Apple apps:", len(apple_data) - unique_apple_apps)
print()
print("Unique Google apps:", unique_google_apps)
print("Total Google apps:", len(google_data))
print("Duplicate Google apps:", len(google_data) - unique_google_apps)

Unique Apple apps: 7195
Total Apple apps: 7197
Duplicate Apple apps: 2

Unique Google apps: 9659
Total Google apps: 10840
Duplicate Google apps: 1181


There are 2 duplicate iOS apps and 1181 duplicate Android apps. Let's start with the apple_data set:

In [19]:
apple_data.track_name.value_counts()[:2]

VR Roller Coaster      2
Mannequin Challenge    2
Name: track_name, dtype: int64

We can see from the above value counts of the track_name column that 'Mannequin Challenge' and 'VR Roller Coaster' are listed twice. Let's look at each one:

In [20]:
apple_data.loc[apple_data['track_name'] == 'Mannequin Challenge']

id           track_name  size_bytes currency  price  \
2948  1173990889  Mannequin Challenge   109705216      USD    0.0   
4463  1178454060  Mannequin Challenge    59572224      USD    0.0   

      rating_count_tot  rating_count_ver  user_rating  user_rating_ver    ver  \
2948               668                87          3.0              3.0    1.4   
4463               105                58          4.0              4.5  1.0.1   

     cont_rating prime_genre  sup_devices.num  ipadSc_urls.num  lang.num  \
2948          9+       Games               37                4         1   
4463          4+       Games               38                5         1   

      vpp_lic  
2948        1  
4463        1

In [21]:
apple_data.loc[apple_data['track_name'] == 'VR Roller Coaster']

id         track_name  size_bytes currency  price  \
4442   952877179  VR Roller Coaster   169523200      USD    0.0   
4831  1089824278  VR Roller Coaster   240964608      USD    0.0   

      rating_count_tot  rating_count_ver  user_rating  user_rating_ver    ver  \
4442               107               102          3.5              3.5  2.0.0   
4831                67                44          3.5              4.0   0.81   

     cont_rating prime_genre  sup_devices.num  ipadSc_urls.num  lang.num  \
4442          4+       Games               37                5         1   
4831          4+       Games               38                0         1   

      vpp_lic  
4442        1  
4831        1

Both apps differ from their duplicate in bytesize, ratings, and in the case of Mannequin Challenge, cont_rating, making them appear to be different apps with the same name. We can leave the apps in the dataset.

Now let's check the Google dataset:

In [22]:
google_data.App.value_counts()[:20]

ROBLOX                                                9
CBS Sports App - Scores, News, Stats & Watch Live     8
Candy Crush Saga                                      7
ESPN                                                  7
Duolingo: Learn Languages Free                        7
8 Ball Pool                                           7
Sniper 3D Gun Shooter: Free Shooting Games - FPS      6
Subway Surfers                                        6
slither.io                                            6
Bleacher Report: sports news, scores, & highlights    6
Bubble Shooter                                        6
Helix Jump                                            6
Bowmasters                                            6
Temple Run 2                                          6
Nick                                                  6
Zombie Catchers                                       6
Zombie Tsunami                                        5
theScore: Live Sports Scores, News, Stats & Vide

There are over 20 duplicates, many of which have 6 or more rows with the same app name. Let's look at the most duplicated app, ROBLOX:

In [23]:
google_data.loc[google_data['App'] == 'ROBLOX']

App Category Rating  Reviews Size      Installs  Type Price  \
1653  ROBLOX     GAME    4.5  4447388  67M  100,000,000+  Free     0   
1701  ROBLOX     GAME    4.5  4447346  67M  100,000,000+  Free     0   
1748  ROBLOX     GAME    4.5  4448791  67M  100,000,000+  Free     0   
1841  ROBLOX     GAME    4.5  4449882  67M  100,000,000+  Free     0   
1870  ROBLOX     GAME    4.5  4449910  67M  100,000,000+  Free     0   
2016  ROBLOX   FAMILY    4.5  4449910  67M  100,000,000+  Free     0   
2088  ROBLOX   FAMILY    4.5  4450855  67M  100,000,000+  Free     0   
2206  ROBLOX   FAMILY    4.5  4450890  67M  100,000,000+  Free     0   
4527  ROBLOX   FAMILY    4.5  4443407  67M  100,000,000+  Free     0   

     Content Rating                        Genres   Last Updated  \
1653   Everyone 10+  Adventure;Action & Adventure  July 31, 2018   
1701   Everyone 10+  Adventure;Action & Adventure  July 31, 2018   
1748   Everyone 10+  Adventure;Action & Adventure  July 31, 2018   
1841   Everyone 10+  Adventure;Action & Adventure  July 31, 2018   
1870   Everyone 10+  Adventure;Action & Adventure  July 31, 2018   
2016   Everyone 10+  Adventure;Action & Adventure  July 31, 2018   
2088   Everyone 10+  Adventure;Action & Adventure  July 31, 2018   
2206   Everyone 10+  Adventure;Action & Adventure  July 31, 2018   
4527   Everyone 10+  Adventure;Action & Adventure  July 31, 2018   

       Current Ver Android Ver  
1653  2.347.225742  4.1 and up  
1701  2.347.225742  4.1 and up  
1748  2.347.225742  4.1 and up  
1841  2.347.225742  4.1 and up  
1870  2.347.225742  4.1 and up  
2016  2.347.225742  4.1 and up  
2088  2.347.225742  4.1 and up  
2206  2.347.225742  4.1 and up  
4527  2.347.225742  4.1 and up

The values in each column look fairly similar, meaning they are likely duplicates. Let's check if the other apps are also duplicates:

In [24]:
google_data.loc[google_data['App'] == 'CBS Sports App - Scores, News, Stats & Watch Live']

App Category Rating  \
2976  CBS Sports App - Scores, News, Stats & Watch Live   SPORTS    4.3   
3007  CBS Sports App - Scores, News, Stats & Watch Live   SPORTS    4.3   
3015  CBS Sports App - Scores, News, Stats & Watch Live   SPORTS    4.3   
3020  CBS Sports App - Scores, News, Stats & Watch Live   SPORTS    4.3   
3056  CBS Sports App - Scores, News, Stats & Watch Live   SPORTS    4.3   
3064  CBS Sports App - Scores, News, Stats & Watch Live   SPORTS    4.3   
3090  CBS Sports App - Scores, News, Stats & Watch Live   SPORTS    4.3   
9594  CBS Sports App - Scores, News, Stats & Watch Live   SPORTS    4.3   

     Reviews                Size    Installs  Type Price Content Rating  \
2976   91031  Varies with device  5,000,000+  Free     0       Everyone   
3007   91031  Varies with device  5,000,000+  Free     0       Everyone   
3015   91031  Varies with device  5,000,000+  Free     0       Everyone   
3020   91031  Varies with device  5,000,000+  Free     0       Everyone   
3056   91033  Varies with device  5,000,000+  Free     0       Everyone   
3064   91033  Varies with device  5,000,000+  Free     0       Everyone   
3090   91033  Varies with device  5,000,000+  Free     0       Everyone   
9594   91035  Varies with device  5,000,000+  Free     0       Everyone   

      Genres    Last Updated         Current Ver Android Ver  
2976  Sports  August 4, 2018  Varies with device  5.0 and up  
3007  Sports  August 4, 2018  Varies with device  5.0 and up  
3015  Sports  August 4, 2018  Varies with device  5.0 and up  
3020  Sports  August 4, 2018  Varies with device  5.0 and up  
3056  Sports  August 4, 2018  Varies with device  5.0 and up  
3064  Sports  August 4, 2018  Varies with device  5.0 and up  
3090  Sports  August 4, 2018  Varies with device  5.0 and up  
9594  Sports  August 4, 2018  Varies with device  5.0 and up

In [25]:
google_data.loc[google_data['App'] == '8 Ball Pool']

App Category Rating   Reviews Size      Installs  Type Price  \
1675  8 Ball Pool     GAME    4.5  14198297  52M  100,000,000+  Free     0   
1703  8 Ball Pool     GAME    4.5  14198602  52M  100,000,000+  Free     0   
1755  8 Ball Pool     GAME    4.5  14200344  52M  100,000,000+  Free     0   
1844  8 Ball Pool     GAME    4.5  14200550  52M  100,000,000+  Free     0   
1871  8 Ball Pool     GAME    4.5  14201891  52M  100,000,000+  Free     0   
1970  8 Ball Pool     GAME    4.5  14201604  52M  100,000,000+  Free     0   
3953  8 Ball Pool   SPORTS    4.5  14184910  52M  100,000,000+  Free     0   

     Content Rating  Genres   Last Updated Current Ver   Android Ver  
1675       Everyone  Sports  July 31, 2018       4.0.0  4.0.3 and up  
1703       Everyone  Sports  July 31, 2018       4.0.0  4.0.3 and up  
1755       Everyone  Sports  July 31, 2018       4.0.0  4.0.3 and up  
1844       Everyone  Sports  July 31, 2018       4.0.0  4.0.3 and up  
1871       Everyone  Sports  July 31, 2018       4.0.0  4.0.3 and up  
1970       Everyone  Sports  July 31, 2018       4.0.0  4.0.3 and up  
3953       Everyone  Sports  July 31, 2018       4.0.0  4.0.3 and up

They all look similar enough to classify as duplicates. We can write a function to iterate through the dataset and remove all these duplicate apps.

#### Deciding which app to keep
We don't want to count certain apps more than once when we analyze data, so we need to remove the duplicate entries and keep only one entry per app. One thing we could do is remove the duplicate rows randomly, but we could probably find a better way.

If you examine the rows we printed above, the main difference happens on the fourth position of each row, which corresponds to the number of reviews. The different numbers show that the data was collected at different times. We can use this to build a criterion for keeping rows. We won't remove rows randomly, but rather we'll keep the rows that have the highest number of reviews because the higher the number of reviews, the more reliable the ratings.

In the above example, we want the 8 Ball Pool game with the following number of reviews:

In [26]:
google_data.loc[google_data['App'] == '8 Ball Pool'].Reviews.max()

'14201891'

First, we'll set the Reviews column as an integer to sort the column:

In [27]:
google_data['Reviews'] = google_data['Reviews'].astype(int)

Now we can use the drop_duplicates function on the column App on the sorted dataframe. This will drop all duplicates values in App name while keeping the last value, which has the highest number of reviews:

In [28]:
google_data = google_data.sort_values('Reviews').drop_duplicates(subset=['App'], keep='last')

In [29]:
google_data

App          Category   Rating  \
6665                           BAR-B-Q Recipes    FOOD_AND_DRINK  Unrated   
7732             SHUTTLLS CQ - Connect Ride Go  TRAVEL_AND_LOCAL  Unrated   
7735                                CQ Ukraine      PRODUCTIVITY  Unrated   
9337                  EG | Explore Folegandros  TRAVEL_AND_LOCAL  Unrated   
7737                       CQ Electrical Group      PRODUCTIVITY  Unrated   
...                                        ...               ...      ...   
1879                            Clash of Clans              GAME      4.6   
382   Messenger – Text and Video Chat for Free     COMMUNICATION        4   
2604                                 Instagram            SOCIAL      4.5   
336                         WhatsApp Messenger     COMMUNICATION      4.4   
2544                                  Facebook            SOCIAL      4.1   

       Reviews                Size        Installs  Type  Price  \
6665         0                3.6M            100+  Free      0   
7732         0                 18M              5+  Free      0   
7735         0                9.1M             10+  Free      0   
9337         0                 56M              0+  Paid  $3.99   
7737         0                 14M              1+  Free      0   
...        ...                 ...             ...   ...    ...   
1879  44893888                 98M    100,000,000+  Free      0   
382   56646578  Varies with device  1,000,000,000+  Free      0   
2604  66577446  Varies with device  1,000,000,000+  Free      0   
336   69119316  Varies with device  1,000,000,000+  Free      0   
2544  78158306  Varies with device  1,000,000,000+  Free      0   

     Content Rating          Genres      Last Updated         Current Ver  \
6665       Everyone    Food & Drink  January 30, 2018                 1.0   
7732       Everyone  Travel & Local    April 26, 2018            4.6.2100   
7735       Everyone    Productivity     June 25, 2018              1.17.0   
9337       Everyone  Travel & Local  January 22, 2017               1.1.1   
7737       Everyone    Productivity     July 19, 2018               1.0.0   
...             ...             ...               ...                 ...   
1879   Everyone 10+        Strategy     July 15, 2018           10.322.16   
382        Everyone   Communication    August 1, 2018  Varies with device   
2604           Teen          Social     July 31, 2018  Varies with device   
336        Everyone   Communication    August 3, 2018  Varies with device   
2544           Teen          Social    August 3, 2018  Varies with device   

             Android Ver  
6665          4.0 and up  
7732          4.3 and up  
7735          4.1 and up  
9337          4.1 and up  
7737          4.1 and up  
...                  ...  
1879          4.1 and up  
382   Varies with device  
2604  Varies with device  
336   Varies with device  
2544  Varies with device  

[9659 rows x 13 columns]

Our google_data dataframe now has the expected 9659 rows with no duplicate values. Let's check the Review count for the duplicate app we listed above:

In [30]:
google_data.loc[google_data['App'] == '8 Ball Pool']

App Category Rating   Reviews Size      Installs  Type Price  \
1871  8 Ball Pool     GAME    4.5  14201891  52M  100,000,000+  Free     0   

     Content Rating  Genres   Last Updated Current Ver   Android Ver  
1871       Everyone  Sports  July 31, 2018       4.0.0  4.0.3 and up

There is only one app under that name, and with the highest number of reviews. We can revert the dataframe to be sorted by index: 

In [31]:
google_data.sort_index(inplace=True)

##  Removing Non-English Apps

If you explore the data sets enough, you'll notice the names of some of the apps suggest they are not directed toward an English-speaking audience. Below, we see a couple of examples from both data sets:

In [32]:
print(apple_data.loc[813].track_name)
print(apple_data.loc[6731].track_name)
print(google_data.loc[5513].App)
print(google_data.loc[9117].App)

爱奇艺PPS -《欢乐颂2》电视剧热播
【脱出ゲーム】絶対に最後までプレイしないで 〜謎解き＆ブロックパズル〜
中国語 AQリスニング
لعبة تقدر تربح DZ


For our data purposes, we're only going to look at Apps in English. So we can create a function to remove all non-English apps from our dataset.

In [33]:
def is_english(s):
    return s.isascii()

In [34]:
print(is_english(apple_data.loc[813].track_name))
print(is_english(apple_data.loc[6731].track_name))
print(is_english(google_data.loc[5513].App))
print(is_english(google_data.loc[9117].App))

False
False
False
False


The `.isascii` functions checks if the characters in a string are ascii. This works well enough to remove apps outside of the alphabet range, but it also removes English apps with non-ascii characters:

In [35]:
print(is_english('Docs To Go™ Free Office Suite'))
print(is_english('Instachat 😜'))

False
False


Instead we can create a new function that will check app names for non-ascii characters and, if there are more than 3, will mark the app as non-English.

In [36]:
def is_english(s):
    non_ascii = 0
    for char in s:
        if ord(char) > 127:
            non_ascii += 1
        if non_ascii > 3:
            return False
    return True

In [37]:
print(is_english('Docs To Go™ Free Office Suite'))
print(is_english('Instachat 😜'))
print(is_english('中国語 AQリスニング'))

True
True
False


This function works well enough for our purposes, although it's possible a few non-English apps might make it past our filter, and some English apps will be removed. But we can work on optimizing this later.

Below, we use the `is_english()` function to filter out the non-English apps for both data sets using the `map` function:

In [38]:
google_data = google_data[google_data['App'].map(is_english)]

In [39]:
apple_data = apple_data[apple_data['track_name'].map(is_english)]

In [40]:
print("Google data rows: ", google_data.shape[0])
print("Apple data rows: ", apple_data.shape[0])

Google data rows:  9614
Apple data rows:  6183


We're now left with 9,614 Google apps and 6,183 Apple apps